sdcflows/transform.py

```diff
@@ -79,47 +79,63 @@ def fit(self, target_reference, affine=None, approx=True):
 
         """
         # Calculate the physical coordinates of target grid
-        if isinstance(spatialimage, (str, bytes, Path)):
-            spatialimage = nb.load(spatialimage)
+        if isinstance(target_reference, (str, bytes, Path)):
+            target_reference = nb.load(target_reference)
 
-        # Initialize xform (or set identity)
-        xfm = self.xfm if self.xfm is not None else nt.linear.Affine()
+        affine = affine if affine is not None else np.eye(4)
+        target_affine = target_reference.affine.copy()
+
+        # Project the reference's grid onto the fieldmap's
+        target_reference = target_reference.__class__(
+            target_reference.dataobj,
+            affine @ target_affine.T,
```

Comment:
Is this right? Since `target_affine` is `IJK -> RAS(target)` and `affine` is `RAS(target) -> RAS(fieldmap)`, I think you want

```suggestion
            affine @ target_affine,
```

Reply:
I believe you are right, but I'm not 100% sure so I opted for adding the `.T` and come back to it later. It really gives me confidence you noticed :).

So why I'm not 100% sure - the affine matrix has 4 column vectors. However, I believe the dot-matrix operator expects the target vectors to be rows. So maybe what's missing is another transposition of the dot product to bring the vectors back to column vectors.

Reply:
Worth revisiting https://en.wikipedia.org/wiki/Affine_transformation to clear your head. I find it easiest to think of affines as a group; the augmented matrix formulation is identical to the group of all invertible affine transformations. A critical property is that the bottom row is `[0, ..., 0, 1]`, which you lose by transposition.

```Python
In [1]: import numpy as np

In [2]: t1 = np.eye(4)

In [3]: t1[:3, 3] = [1, 2, 3]

In [4]: t2 = 2 * np.eye(4)

In [5]: t2[:, 3] = [4, 5, 6, 1]

In [6]: t1 @ t2
Out[6]: 
array([[2., 0., 0., 5.],
       [0., 2., 0., 7.],
       [0., 0., 2., 9.],
       [0., 0., 0., 1.]])

In [7]: t1 @ t2.T
Out[7]: 
array([[ 6.,  5.,  6.,  1.],
       [ 8., 12., 12.,  2.],
       [12., 15., 20.,  3.],
       [ 4.,  5.,  6.,  1.]])
```

Bearing in mind the group structure helps keep straight that there are only two things you can do: multiply (with the correct product) and invert. Now all you need to keep track of is that the source vector space is on the right and the destination on the left.